src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,227 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
+                _sources = sources;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNIterator<TSource>(_sources);
+            }
+
+            protected override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                return index < _sources.Length ? _sources[index] : null;
+            }
+        }
+
+        private abstract class ConcatIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        {
+            private IEnumerator<TSource> _enumerator;
+
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                }
+                base.Dispose();
+            }
+
+            protected abstract IEnumerable<TSource> GetEnumerable(int index);
+
+            public override bool MoveNext()
+            {
+                if (state == 1)
+                {
+                    _enumerator = GetEnumerable(0).GetEnumerator();
+                    state = 2;
+                }
+
+                if (state > 1)
+                {
+                    while (true)
+                    {
+                        if (_enumerator.MoveNext())
+                        {
+                            current = _enumerator.Current;
+                            return true;
+                        }
+
+                        Debug.Assert(state < int.MaxValue);
+                        IEnumerable<TSource> next = GetEnumerable(state++ - 1);
+                        if (next != null)
+                        {
+                            _enumerator.Dispose();
+                            _enumerator = next.GetEnumerator();
+                            continue;
+                        }
+
+                        Dispose();
+                        break;
+                    }
+                }
+
+                return false;
+            }
+
+            public TSource[] ToArray()
+            {
+                return EnumerableHelpers.ToArray(this);
+            }
+
+            public List<TSource> ToList()
+            {
+                var list = new List<TSource>();
+                for (int i = 0; ; i++)
+                {
+                    IEnumerable<TSource> source = GetEnumerable(i);
+                    if (source == null) break;
+
+                    list.AddRange(source);
+                }
+                return list;
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                int count = 0;
+                for (int i = 0; ; i++)
+                {
+                    IEnumerable<TSource> source = GetEnumerable(i);
+                    if (source == null) break;
+
+                    ICollection<TSource> c = source as ICollection<TSource>;
```

Comment:
Honestly, I was actually hesitant to even take it this far... I'm a bit concerned that having to look at all of the constituent enumerables, do all of the casting and type checks, etc., could actually hurt cases where onlyIfCheap and it needs to be really, really cheap.  I think I may just scale this back to always return -1 if onlyIfCheap.